bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java

```diff
@@ -17,377 +17,21 @@
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf getBytes(int index, OutputStream out, int length) throws IOException {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setByte(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    protected void _setByte(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setShort(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    protected void _setShort(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setMedium(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    protected void _setMedium(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setInt(int index, int value) {
-        return (DoubleByteBuf) super.setInt(index, value);
-    }
-
-    @Override
-    protected void _setInt(int index, int value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setLong(int index, long value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    protected void _setLong(int index, long value) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setBytes(int index, byte[] src, int srcIndex, int length) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setBytes(int index, ByteBuffer src) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public DoubleByteBuf setBytes(int index, ByteBuf src, int srcIndex, int length) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public int setBytes(int index, InputStream in, int length) throws IOException {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public int setBytes(int index, ScatteringByteChannel in, int length) throws IOException {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public ByteBuf copy(int index, int length) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public int nioBufferCount() {
-        return b1.nioBufferCount() + b2.nioBufferCount();
-    }
-
-    @Override
-    public ByteBuffer internalNioBuffer(int index, int length) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public ByteBuffer nioBuffer(int index, int length) {
-        ByteBuffer dst = isDirect() ? ByteBuffer.allocateDirect(length) : ByteBuffer.allocate(length);
-        ByteBuf b = Unpooled.wrappedBuffer(dst);
```

Comment:
so, is this the cause for allocating unpooled direct ByteBuffer?

Reply:
Yes, until netty-4.1.12 it was going through the `DoubleByteBuf.getBytes()` and so there were no allocations.

Reply:
Do we have the option to use older netty in bookie? In that case, we need to keep old code as well.

Reply:
@saandrews @rdhabalia  It's not possible to use get one implementation or the other depending on the Netty version at runtime. 

I'm still testing on a fix to get rid of DoubleByteBuf/CompositeByteBuf, but that's not a trivial change, I cannot get it working reliably yet.

I haven't seen any performance degradation going from Pulsar-1.21 to the current master with this patch applied. If you guys are worried, you can still depend on the 4.3.1.85-yahoo BK artifact for the internal deployment, until we merge the final fix.

Right now, the current implementation degrades badly with newer Netty, so we need a quick fix for it.

Reply:
I see. Lets proceed with this for now then.